modules/core/include/opencv2/core/hal/intrin_math.hpp

```diff
@@ -34,434 +30,428 @@ namespace CV__SIMD_NAMESPACE {
-        const v_float16 _vexp_p4_f16 = vx_setall_f16(1.6666665459E-1f);
-        const v_float16 _vexp_p5_f16 = vx_setall_f16(5.0000001201E-1f);
-        const v_int16 _vexp_bias_s16 = vx_setall_s16(0xf);
-
-        v_float16 _vexp_, _vexp_x, _vexp_y, _vexp_xx;
-        v_int16 _vexp_mm;
-
-        // compute exponential of x
-        _vexp_x = v_max(x, _vexp_lo_f16);
-        _vexp_x = v_min(_vexp_x, _vexp_hi_f16);
-
-        _vexp_ = v_fma(_vexp_x, _vexp_LOG2EF_f16, _vexp_half_fp16);
-        _vexp_mm = v_floor(_vexp_);
-        _vexp_ = v_cvt_f16(_vexp_mm);
-        _vexp_mm = v_add(_vexp_mm, _vexp_bias_s16);
-        _vexp_mm = v_shl(_vexp_mm, 10);
-
-        _vexp_x = v_fma(_vexp_, _vexp_C1_f16, _vexp_x);
-        _vexp_x = v_fma(_vexp_, _vexp_C2_f16, _vexp_x);
-        _vexp_xx = v_mul(_vexp_x, _vexp_x);
-
-        _vexp_y = v_fma(_vexp_x, _vexp_p0_f16, _vexp_p1_f16);
-        _vexp_y = v_fma(_vexp_y, _vexp_x, _vexp_p2_f16);
-        _vexp_y = v_fma(_vexp_y, _vexp_x, _vexp_p3_f16);
-        _vexp_y = v_fma(_vexp_y, _vexp_x, _vexp_p4_f16);
-        _vexp_y = v_fma(_vexp_y, _vexp_x, _vexp_p5_f16);
-
-        _vexp_y = v_fma(_vexp_y, _vexp_xx, _vexp_x);
-        _vexp_y = v_add(_vexp_y, _vexp_one_fp16);
-        _vexp_y = v_mul(_vexp_y, v_reinterpret_as_f16(_vexp_mm));
-
-        // exp(NAN) -> NAN
-        v_float16 mask_not_nan = v_not_nan(x);
-        return v_select(mask_not_nan, _vexp_y, v_reinterpret_as_f16(vx_setall_s16(0x7e00)));
+// Implementation is the same as float32 vector.
+#define OPENCV_HAL_MATH_IMPL_EXP_F16(VPrefix, TpSuffix) \
+inline v_float##TpSuffix v_exp(const v_float##TpSuffix &x){ \
+        const v_float##TpSuffix _vexp_lo_f16 = VPrefix##_setall_f16(-10.7421875f); \
+        const v_float##TpSuffix _vexp_hi_f16 = VPrefix##_setall_f16(11.f); \
+        const v_float##TpSuffix _vexp_half_fp16 = VPrefix##_setall_f16(0.5f); \
+        const v_float##TpSuffix _vexp_one_fp16 = VPrefix##_setall_f16(1.f); \
+        const v_float##TpSuffix _vexp_LOG2EF_f16 = VPrefix##_setall_f16(1.44269504088896341f); \
+        const v_float##TpSuffix _vexp_C1_f16 = VPrefix##_setall_f16(-6.93359375E-1f); \
+        const v_float##TpSuffix _vexp_C2_f16 = VPrefix##_setall_f16(2.12194440E-4f); \
+        const v_float##TpSuffix _vexp_p0_f16 = VPrefix##_setall_f16(1.9875691500E-4f); \
+        const v_float##TpSuffix _vexp_p1_f16 = VPrefix##_setall_f16(1.3981999507E-3f); \
+        const v_float##TpSuffix _vexp_p2_f16 = VPrefix##_setall_f16(8.3334519073E-3f); \
+        const v_float##TpSuffix _vexp_p3_f16 = VPrefix##_setall_f16(4.1665795894E-2f); \
+        const v_float##TpSuffix _vexp_p4_f16 = VPrefix##_setall_f16(1.6666665459E-1f); \
+        const v_float##TpSuffix _vexp_p5_f16 = VPrefix##_setall_f16(5.0000001201E-1f); \
+        const v_int##TpSuffix _vexp_bias_s16 = VPrefix##_setall_s16(0xf); \
+\
+        v_float##TpSuffix _vexp_, _vexp_x, _vexp_y, _vexp_xx; \
+        v_int##TpSuffix _vexp_mm; \
```

Comment:
I propose to use templates here. v_int##TpSuffix and v_float##TpSuffix will be template arguments.

Reply:
OK, but it will introduce many parameters and may use the macro to maintain the code.

Reply:
@asmorkalov How is `VPrefix` (e.g., v256_setall_f32, v512_setall_f32) set depending on the platform? The function implementation cannot utilize templates.

Reply:
I see. Thanks a lot for the investigation!

Reply:
Wouldn't wide intrinsics like `vx_setall_f32` work here? Or maybe we can introduce traits type (or extend existing ones) which would map `_setall_` to corresponding types.

Reply:
@mshabunin The code will generate v_func for all backends. So when the user passes the widest vector variable, it will call the widest implementation version. (vx_setall_f32 will set the widest vector)